File: golang/compute_heavy.go
package main

import (
	"fmt"
	"os"
	"time"
)

func matrixDeterminant(size uint64) uint64 {
	n := size + 2
	matrix := make([][]uint64, n)
	for i := range matrix {
		matrix[i] = make([]uint64, n)
	}

	for i := uint64(0); i < n; i++ {
		for j := uint64(0); j < n; j++ {
			matrix[i][j] = (i*j + 1) % 1000
		}
	}

	det := uint64(1)
	for i := uint64(0); i < n; i++ {
		for j := uint64(0); j < n; j++ {
			det = (det * matrix[i][j]) % 1000000
		}
	}
	return det
}

func factorial(n uint64) uint64 {
	if n == 0 || n == 1 {
		return 1
	}
	result := uint64(1)
	for i := uint64(2); i <= n; i++ {
		result = result * i
	}
	return result
}

func hashCompute(seed uint64) uint64 {
	hash := seed
	for i := 0; i < 100; i++ {
		hash = hash*1103515245 + 12345
		hash = (hash >> 16) ^ hash
	}
	return hash
}

func main() {
	start := time.Now()

	result := uint64(0)
	iterations := uint64(1_000_000)

	for i := uint64(0); i < iterations; i++ {
		result += matrixDeterminant(i % 15)
		result += factorial(i % 18)
		result += hashCompute(i)
	}

	duration := time.Since(start).Nanoseconds()
	fmt.Println(duration)
	fmt.Printf("Result: %d\n", result)
	os.Exit(0)
}

File: python/compute_heavy.py
import time

def matrix_determinant(size):
    n = size + 2
    matrix = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(n):
        for j in range(n):
            matrix[i][j] = (i * j + 1) % 1000

    det = 1
    for i in range(n):
        for j in range(n):
            det = (det * matrix[i][j]) % 1000000
    return det

def factorial(n):
    if n == 0 or n == 1:
        return 1
    result = 1
    for i in range(2, n + 1):
        result = (result * i) & 0xFFFFFFFFFFFFFFFF
    return result

def hash_compute(seed):
    hash_val = seed
    for _ in range(100):
        hash_val = ((hash_val * 1103515245 + 12345) & 0xFFFFFFFFFFFFFFFF)
        hash_val = ((hash_val >> 16) ^ hash_val) & 0xFFFFFFFFFFFFFFFF
    return hash_val

def main():
    start = time.time_ns()

    result = 0
    iterations = 1_000_000

    for i in range(iterations):
        result = (result + matrix_determinant(i % 15)) & 0xFFFFFFFFFFFFFFFF
        result = (result + factorial(i % 18)) & 0xFFFFFFFFFFFFFFFF
        result = (result + hash_compute(i)) & 0xFFFFFFFFFFFFFFFF

    duration = time.time_ns() - start
    print(duration)
    print(f"Result: {result}")
    exit(0)

if __name__ == "__main__":
    main()
